internal/goldendataset/span_generator.go

```diff
@@ -0,0 +1,559 @@
+	statusMap[SpanStatusUnavailable] = "RecommendationService is currently unavailable"
+	statusMap[SpanStatusDataLoss] = ""
+	statusMap[SpanStatusUnauthenticated] = "nstark is unknown user"
+	return statusMap
+}
+
+func GenerateSpans(count int, startPos int, pictFile string) ([]*otlptrace.Span, int, error) {
+	pairsData, err := loadPictOutputFile(pictFile)
+	if err != nil {
+		return nil, 0, err
+	}
+	pairsTotal := len(pairsData)
+	spanList := make([]*otlptrace.Span, count)
+	index := startPos + 1
+	var inputs []string
+	traceID := generateTraceID()
+	parentID := generateSpanID()
+	for i := 0; i < count; i++ {
+		inputs = pairsData[index]
+		if SpanParentRoot == inputs[ColumnParent] {
+			traceID = generateTraceID()
+			parentID = nil
+		}
+		spanName := generateSpanName(inputs[ColumnAttributes], i)
+		spanList[i] = GenerateSpan(traceID, inputs[ColumnTracestate], parentID, spanName, inputs[ColumnKind],
+			inputs[ColumnAttributes], inputs[ColumnEvents], inputs[ColumnLinks], inputs[ColumnStatus])
+		parentID = spanList[i].SpanId
+		index++
+		if index >= pairsTotal {
+			index = 1
+		}
+	}
+	return spanList, index, nil
+}
+
+func loadPictOutputFile(fileName string) ([][]string, error) {
+	file, err := os.Open(filepath.Clean(fileName))
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		cerr := file.Close()
+		if err == nil {
+			err = cerr
+		}
+	}()
+
+	reader := csv.NewReader(file)
+	reader.Comma = '\t'
+
+	return reader.ReadAll()
+}
+
+func generateSpanName(spanTypeID string, index int) string {
+	if SpanAttrHTTPClient == spanTypeID {
+		return fmt.Sprintf("/dragons/%d", index)
+	} else if SpanAttrHTTPServer == spanTypeID {
+		return "/dragons/{dragonId}"
+	} else if SpanAttrGRPCClient == spanTypeID || SpanAttrGRPCServer == spanTypeID {
+		return "com.example.PetFoodService/DispenseFeed"
+	} else {
+		return fmt.Sprintf("gotest%d", index)
+	}
+}
+
+func GenerateSpan(traceID []byte, tracestate string, parentID []byte, spanName string, kind string, spanTypeID string,
+	eventCnt string, linkCnt string, statusStr string) *otlptrace.Span {
+	endTime := time.Now().Add(-50 * time.Microsecond)
+	return &otlptrace.Span{
+		TraceId:                traceID,
+		SpanId:                 generateSpanID(),
+		TraceState:             generateTraceState(tracestate),
+		ParentSpanId:           parentID,
+		Name:                   spanName,
+		Kind:                   lookupSpanKind(kind),
+		StartTimeUnixNano:      uint64(endTime.Add(-215 * time.Millisecond).UnixNano()),
+		EndTimeUnixNano:        uint64(endTime.UnixNano()),
+		Attributes:             generateSpanAttributes(spanTypeID),
+		DroppedAttributesCount: 0,
+		Events:                 generateSpanEvents(eventCnt),
+		DroppedEventsCount:     0,
+		Links:                  generateSpanLinks(linkCnt),
+		DroppedLinksCount:      0,
+		Status:                 generateStatus(statusStr),
+	}
+}
+
+func generateTraceID() []byte {
+	var r [16]byte
+	_, err := rand.Read(r[:])
```

Comment:
Do we want the generator to be trully pseudo-random?
It may be very useful to instead pass an instance of rand.Rand to the generator. This will allow to pre-seed the random generator with the same value in comparison tests that use the generator. The seed may still be random for each execution of the test suite but it is important that comparison tests use the same seed and same sequence of random data so that comparison is valid.

This will require to make the generator a statefull object where Rand instance can live but I think it will make the generator more widely useful.

This is important for example for tests which can exhibit different results depending on the content of data, e.g. benchmarks that calculate compression ratio of different compression algorithms for the same input data.

Reply:
Changed so the generator is passed in from top-level generation method calls